src/plan/plan_constraints.rs

```diff
@@ -1,20 +1,40 @@
-// Specific plans may 'pub use' the constraints here, and overwrites some. In that case,
-// the constants that get overwritten are unused.
-#![allow(unused)]
-
 use crate::util::constants::*;
 
-pub const MOVES_OBJECTS: bool = false;
-pub const NUM_SPECIALIZED_SCANS: usize = 0;
-// The two consts below need to be consistent
-pub const NEEDS_LOG_BIT_IN_HEADER: bool = false;
-pub const NEEDS_LOG_BIT_IN_HEADER_NUM: usize = 0;
-
-pub const NEEDS_LINEAR_SCAN: bool = SUPPORT_CARD_SCANNING || LAZY_SWEEP;
-pub const NEEDS_CONCURRENT_WORKERS: bool = false;
-
-pub const GENERATE_GC_TRACE: bool = false;
-
-pub const MAX_NON_LOS_COPY_BYTES: usize = MAX_INT;
+/// This struct defines plan-specific constraints.
+/// Most of the constraints are constants. Each plan should declare a constant of this struct,
+/// and use the constant wherever possible. However, for plan-neutral implementations,
+/// these constraints are not constant.
+pub struct PlanConstraints {
```

Comment:
Is it possible we have another `PlanConstraintsTrait` like the following

```rust
trait PlanConstraintsTrait {
    const MOVES_OBJECTS: bool;
    const GC_HEADER_BITS: usize;
    const GC_HEADER_WORDS: usize;
    // ... other constants
    fn constraints() -> PlanConstraints {
        PlanConstraints {
            moves_objects: Self::MOVE_OBJECTS,
            gc_header_bits: Self::GC_HEADER_BITS,
            gc_header_words: Self::GC_HEADER_WORDS,
            // ... other constants
        }
    }
}
```

And then say, for a GenCopy plan:

```rust
impl PlanConstraintsTrait for GenCopy {
    const MOVES_OBJECTS: bool = false;
    const GC_HEADER_BITS: usize = 0;
    const GC_HEADER_WORDS: usize = 0;
    // ... other constants
}

impl Plan for GenCopy {
    fn constraints(&self) -> PlanConstraints {
        <Self as PlanConstraintsTrait>::constraints()
    }
}
```

Although we're now having two interfaces for accessing these constants, as long as some code can known the type of a plan, it can guarantee to inline these constants if they are using `const` defined constants.

Probably `fn constraints` is also possible to be inlined properly. So this may not be a problem...

Reply:
This should be possible. What is the benefit of doing this?

Reply:
Just to ensure the code can inline these constants if possible.
Probably it may not worth doing so... I think if the specific type of a `Plan` is known at compile time, the compiler can also inline the `fn constraints(&self)` call, and eventually inline those values defined in `const GENCOPY_CONSTRAINTS: PlanConstraints`.

Reply:
For the plan-specific code, they can always use the constant like `GENCOPY_CONSTRAINTS` - I think the compiler can do the right job. For plan-neutral code, I don't expect they can use plan constraints as a constant. For `plan.constraints()` where `plan` is a `dyn Plan`, I *assume* there is always a dynamic dispatch. It is not about the implementation of `constraints()`, it is the `dyn Plan`. 

Do you think we can improve on this with your proposal?